keps/sig-node/4205-psi-metric/README.md

```diff
@@ -0,0 +1,646 @@
+
+In short, PSI metric are like barometers that provide fair warning of impending resource shortages on the node, and enable nodes to take more proactive, granular and nuanced steps when major resources (memory, CPU, IO) start becoming scarce.
+
+### Goals
+
+This proposal aims to:
+1. Enable the kubelet to have the PSI metric of cgroupv2 exposed from cAdvisor and Runc.
+2. Enable the pod level PSI metric and expose it in the Summary API.
+3. Utilize the node level PSI metric to set node condition and node taints.
+
+It will have two phases:
+Phase 1: includes goal 1, 2 
+Phase 2: includes goal 3
+
+### Non-Goals
+
+* Invest in more opportunities to further use PSI metric for pod evictions,
+userspace OOM kills, and so on, for future KEPs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+Today, to identity disruptions caused by resource crunches, Kubernetes users need to
+install node exporter to read PSI metric. With the feature proposed in this enhancement, 
+PSI metric will be available for customers in the Kubernetes metrics API.
+
+#### Story 2
+
+Kubernetes users want to prevent new pods to be scheduled on the nodes that have resource starvation. By using PSI metric, the kubelet will set Node Condition to avoid pods being scheduled on nodes under high resource pressure. The node controller could then set a (taint on the node based on these new Node Conditions)[https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-nodes-by-condition].
+
+### Risks and Mitigations
+
+There are no significant risks associated with Phase 1 implementation that involves integrating
+the PSI metric in kubelet from runc, and exposing these through the metric endpoint.
+
+Phase 2 involves utilizing the PSI metric to report node conditions. There is a potential
+risk of early reporting for nodes under pressure. We intend to address this concern
+by conducting careful experimentation with PSI threshold values to identify the optimal
+default threshold to be used for reporting the nodes under heavy resource pressure.
+
+## Design Details
+
+#### Phase 1
+1. Add new Data structures PSIData and PSIStats corresponding to the PSI metric output format as following:
+some avg10=0.00 avg60=0.00 avg300=0.00 total=0
+full avg10=0.00 avg60=0.00 avg300=0.00 total=0
+
+```go
+type PSIData struct {
+	Avg10 *float64 `json:”avg10”`
+	Avg60 *float64 `json:”avg60”`
+Avg300 *float64 `json:”avg300”`
+Total *float64 `json:”total”`
+}
+
+type PSIStats struct {
+	Some *PSIData  `json:”some,omitempty”`
+Full *PSIData  `json:”full,omitempty”`
+}
+```
+
+2. Summary API includes stats for both system and kubepods level cgroups. Extend the Summary API to include PSI metric data for each resource obtained from cadvisor. 
```

Comment:
I think it may be worth to go into a bit more detail here that the summary API currently exposes a concept of `.node.systemContainers` with 3 system containers, `pods`, `kubelet`, and `runtime` -- https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/summary_sys_containers.go#L71-L74

Additionally, the root cgroup stats are reported as part of node.{cpu,memory}, etc -- https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/summary.go#L143-L146

Do we plan add PSI stats for each of those (i.e. each of the system containers and the root stats as well)?

Reply:
This is related to @rphillips comment regarding collecting PSI for each of the main slices. @rphillips is this what you had in mind as well?